File: django_currency_converter/services/currency_converter.py
import requests
import logging
from datetime import datetime, date, timedelta
from decimal import Decimal, InvalidOperation
from typing import Optional, Dict, List, Tuple
from django.conf import settings
from django.core.cache import cache
from django.utils import timezone
from django.db import transaction as db_transaction
from ..models import ExchangeRateCache, SupportedCurrency

logger = logging.getLogger(__name__)


class CurrencyConversionError(Exception):
    """Custom exception for currency conversion errors"""
    pass


class ExchangeRateAPIError(Exception):
    """Exception for API-related errors"""
    pass


class CurrencyConverter:
    """
    Historical currency converter using ExchangeRate.host API
    Supports caching and handles rate limits gracefully
    """
    
    # Alternative free APIs that don't require keys
    BASE_URL = "https://api.exchangerate-api.com/v4"  # Free tier available
    FALLBACK_URL = "https://api.fxratesapi.com"  # Another free option
    CACHE_PREFIX = "exchange_rate"
    CACHE_TIMEOUT = 86400  # 24 hours in seconds
    API_TIMEOUT = 10  # seconds
    MAX_RETRIES = 3
    
    # Common currencies supported by most free APIs
    DEFAULT_SUPPORTED_CURRENCIES = [
        'USD', 'EUR', 'GBP', 'JPY', 'AUD', 'CAD', 'CHF', 'CNY', 'SEK', 'NZD',
        'MXN', 'SGD', 'HKD', 'NOK', 'TRY', 'RUB', 'INR', 'BRL', 'ZAR', 'KRW',
        'PLN', 'CZK', 'HUF', 'RON', 'BGN', 'HRK', 'DKK', 'THB', 'MYR', 'PHP',
        'IDR', 'VND', 'EGP', 'MAD', 'NGN', 'KES', 'GHS', 'UGX', 'TZS', 'ZMW'
    ]

    def __init__(self):
        self.session = requests.Session()
        self.session.timeout = self.API_TIMEOUT

    def get_supported_currencies(self, refresh=False) -> List[str]:
        """
        Get list of supported currencies from API or cache
        
        Args:
            refresh: If True, refresh from API. Otherwise use cached/DB values.
            
        Returns:
            List of supported currency codes
        """
        cache_key = f"{self.CACHE_PREFIX}_supported_currencies"
        
        if not refresh:
            # Try cache first
            cached_currencies = cache.get(cache_key)
            if cached_currencies:
                return cached_currencies
            
            # Try database
            db_currencies = SupportedCurrency.get_supported_codes()
            if db_currencies:
                cache.set(cache_key, db_currencies, self.CACHE_TIMEOUT)
                return db_currencies
        
        # Fetch from API
        try:
            # Use exchangerate-api.com which is free and doesn't require key
            response = self.session.get(f"{self.BASE_URL}/latest/USD")
            response.raise_for_status()
            
            data = response.json()
            if 'rates' not in data:
                raise ExchangeRateAPIError(f"API error: Invalid response format")
            
            api_currencies = list(data.get('rates', {}).keys())
            # Add USD as base currency
            api_currencies.append('USD')
            
            # Update database with supported currencies
            self._update_supported_currencies(api_currencies)
            
            # Cache the result
            cache.set(cache_key, api_currencies, self.CACHE_TIMEOUT)
            
            logger.info(f"Fetched {len(api_currencies)} supported currencies from API")
            return api_currencies
            
        except Exception as e:
            logger.warning(f"Failed to fetch supported currencies from API: {e}")
            # Fallback to default list
            self._update_supported_currencies(self.DEFAULT_SUPPORTED_CURRENCIES)
            cache.set(cache_key, self.DEFAULT_SUPPORTED_CURRENCIES, self.CACHE_TIMEOUT)
            return self.DEFAULT_SUPPORTED_CURRENCIES

    def _update_supported_currencies(self, currency_codes: List[str]):
        """Update the SupportedCurrency model with API results"""
        with db_transaction.atomic():
            # Mark all as unsupported first
            SupportedCurrency.objects.all().update(is_supported=False)
            
            # Update/create supported currencies
            for code in currency_codes:
                SupportedCurrency.objects.update_or_create(
                    code=code.upper(),
                    defaults={
                        'is_supported': True,
                        'name': self._get_currency_name(code)
                    }
                )

    def _get_currency_name(self, code: str) -> str:
        """Get human-readable currency name"""
        currency_names = {
            'USD': 'US Dollar', 'EUR': 'Euro', 'GBP': 'British Pound',
            'JPY': 'Japanese Yen', 'AUD': 'Australian Dollar', 'CAD': 'Canadian Dollar',
            'CHF': 'Swiss Franc', 'CNY': 'Chinese Yuan', 'SEK': 'Swedish Krona',
            'NZD': 'New Zealand Dollar', 'MXN': 'Mexican Peso', 'SGD': 'Singapore Dollar',
            'HKD': 'Hong Kong Dollar', 'NOK': 'Norwegian Krone', 'TRY': 'Turkish Lira',
            'RUB': 'Russian Ruble', 'INR': 'Indian Rupee', 'BRL': 'Brazilian Real',
            'ZAR': 'South African Rand', 'KRW': 'South Korean Won', 'MMK': 'Myanmar Kyat'
        }
        return currency_names.get(code.upper(), f"{code.upper()} Currency")

    def is_currency_supported(self, currency_code: str) -> bool:
        """Check if a currency is supported for conversion"""
        if currency_code.upper() == 'USD':
            return True
        
        supported_currencies = self.get_supported_currencies()
        return currency_code.upper() in supported_currencies

    def get_exchange_rate(self, from_currency: str, to_currency: str, date_value: date) -> Optional[Decimal]:
        """
        Get historical exchange rate for a specific date
        
        Args:
            from_currency: Source currency code
            to_currency: Target currency code (should be USD)
            date_value: Date for the exchange rate
            
        Returns:
            Exchange rate as Decimal, or None if not available
        """
        from_currency = from_currency.upper()
        to_currency = to_currency.upper()
        
        # Handle USD to USD
        if from_currency == to_currency:
            return Decimal('1.0')
        
        # Check if currency is supported
        if not self.is_currency_supported(from_currency):
            logger.warning(f"Currency {from_currency} is not supported")
            return None
        
        # Check cache first (DB cache)
        try:
            cached_rate = ExchangeRateCache.objects.get(
                currency=from_currency,
                date=date_value
            )
            logger.debug(f"Using cached rate for {from_currency} on {date_value}: {cached_rate.rate_to_usd}")
            return cached_rate.rate_to_usd
        except ExchangeRateCache.DoesNotExist:
            pass
        
        # Check memory cache
        cache_key = f"{self.CACHE_PREFIX}_{from_currency}_{to_currency}_{date_value}"
        cached_rate = cache.get(cache_key)
        if cached_rate:
            logger.debug(f"Using memory cached rate for {from_currency} on {date_value}: {cached_rate}")
            return Decimal(str(cached_rate))
        
        # Fetch from API
        return self._fetch_rate_from_api(from_currency, to_currency, date_value)

    def _fetch_rate_from_api(self, from_currency: str, to_currency: str, date_value: date) -> Optional[Decimal]:
        """Fetch exchange rate from ExchangeRate.host API"""
        
        # Don't fetch rates for future dates
        if date_value > date.today():
            logger.warning(f"Cannot fetch exchange rate for future date: {date_value}")
            return None
        
        # Format date for API
        date_str = date_value.strftime('%Y-%m-%d')
        
        for attempt in range(self.MAX_RETRIES):
            try:
                # Use exchangerate-api.com historical endpoint
                # For historical data, try different approach
                if from_currency == 'USD':
                    # Direct USD to other currency
                    url = f"{self.BASE_URL}/history/{date_str}"
                    response = self.session.get(url)
                else:
                    # Get current rate (free APIs often don't have historical data)
                    # For production, you'd need a paid API for historical rates
                    url = f"{self.BASE_URL}/latest/{from_currency}"
                    response = self.session.get(url)
                
                logger.debug(f"Fetching rate from API: {url}")
                
                response.raise_for_status()
                data = response.json()
                
                rates = data.get('rates', {})
                if to_currency not in rates:
                    logger.error(f"Rate for {to_currency} not found in API response")
                    return None
                
                rate = Decimal(str(rates[to_currency]))
                
                # Cache the result in memory
                cache_key = f"{self.CACHE_PREFIX}_{from_currency}_{to_currency}_{date_value}"
                cache.set(cache_key, float(rate), self.CACHE_TIMEOUT)
                
                # Cache in database
                ExchangeRateCache.objects.update_or_create(
                    currency=from_currency,
                    date=date_value,
                    defaults={'rate_to_usd': rate}
                )
                
                logger.info(f"Fetched rate {from_currency} -> {to_currency} on {date_value}: {rate}")
                return rate
                
            except requests.exceptions.RequestException as e:
                logger.warning(f"API request failed (attempt {attempt + 1}/{self.MAX_RETRIES}): {e}")
                if attempt < self.MAX_RETRIES - 1:
                    # Wait before retry (exponential backoff)
                    import time
                    time.sleep(2 ** attempt)
                continue
                
            except (ValueError, InvalidOperation, KeyError) as e:
                logger.error(f"Error parsing API response: {e}")
                return None
        
        logger.error(f"Failed to fetch exchange rate after {self.MAX_RETRIES} attempts")
        return None

    def convert_to_usd(self, amount: Decimal, currency: str, transaction_date: date) -> Tuple[Optional[Decimal], Optional[Decimal]]:
        """
        Convert amount to USD using historical exchange rate
        
        Args:
            amount: Amount to convert
            currency: Currency code of the amount
            transaction_date: Date for historical rate lookup
            
        Returns:
            Tuple of (usd_amount, exchange_rate_used) or (None, None) if conversion failed
        """
        if not amount or amount <= 0:
            return None, None
        
        currency = currency.upper()
        
        # Already USD
        if currency == 'USD':
            return amount, Decimal('1.0')
        
        # Get exchange rate
        exchange_rate = self.get_exchange_rate(currency, 'USD', transaction_date)
        if exchange_rate is None:
            return None, None
        
        try:
            usd_amount = amount * exchange_rate
            # Round to 2 decimal places for currency
            usd_amount = usd_amount.quantize(Decimal('0.01'))
            
            logger.debug(f"Converted {amount} {currency} to {usd_amount} USD using rate {exchange_rate}")
            return usd_amount, exchange_rate
            
        except (InvalidOperation, OverflowError) as e:
            logger.error(f"Error converting {amount} {currency} to USD: {e}")
            return None, None

# Global converter instance
converter = CurrencyConverter()

File: frontend/src/lib/currency-converter.ts
/**
 * Enhanced Historical Currency Converter Service for AIMS
 * Integrates with ExchangeRate.host for reliable historical currency conversion
 * Uses Supabase for intelligent caching and supported currency management
 */

import { getSupabaseAdmin } from './supabase';

export interface ExchangeRate {
  id?: string;
  from_currency: string;
  to_currency: string;
  exchange_rate: number;
  rate_date: string;
  source: string;
  created_at?: string;
}

export interface SupportedCurrency {
  id?: string;
  code: string;
  name: string;
  is_supported: boolean;
  last_checked?: string;
  created_at?: string;
}

export interface ConversionResult {
  usd_amount: number | null;
  exchange_rate: number | null;
  success: boolean;
  error?: string;
  source?: string;
  conversion_date?: string;
}

export interface ConversionStats {
  total_transactions: number;
  converted_transactions: number;
  unconvertible_transactions: number;
  pending_transactions: number;
  usd_transactions: number;
  conversion_rate: number;
}

class EnhancedCurrencyConverter {
  private readonly API_BASE_URL = 'https://api.exchangerate.host';
  private readonly CACHE_DURATION = 7 * 24 * 60 * 60 * 1000; // 7 days in milliseconds
  private readonly MAX_RETRIES = 3;
  private readonly RETRY_DELAY = 1000; // 1 second
  
  // Comprehensive list of supported currencies
  private readonly DEFAULT_CURRENCIES = [
    'USD', 'EUR', 'GBP', 'JPY', 'AUD', 'CAD', 'CHF', 'CNY', 'SEK', 'NZD',
    'MXN', 'SGD', 'HKD', 'NOK', 'TRY', 'RUB', 'INR', 'BRL', 'ZAR', 'KRW',
    'PLN', 'CZK', 'HUF', 'RON', 'BGN', 'HRK', 'DKK', 'THB', 'MYR', 'PHP',
    'IDR', 'VND', 'CLP', 'PEN', 'COP', 'ARS', 'UYU', 'BOB', 'PYG', 'GHS',
    'NGN', 'KES', 'UGX', 'TZS', 'RWF', 'ETB', 'EGP', 'MAD', 'TND', 'DZD',
    'LKR', 'PKR', 'BDT', 'NPR', 'BTN', 'MVR', 'AFN', 'IRR', 'IQD', 'JOD',
    'KWD', 'BHD', 'QAR', 'AED', 'SAR', 'OMR', 'YER', 'LBP', 'SYP', 'ILS'
  ];

  /**
   * Get historical exchange rate using ExchangeRate.host
   */
  async getHistoricalRate(
    fromCurrency: string, 
    toCurrency: string, 
    date: Date
  ): Promise<{ rate: number; source: string } | null> {
    const from = fromCurrency.toUpperCase();
    const to = toCurrency.toUpperCase();
    const dateStr = date.toISOString().split('T')[0];

    // Handle same currency conversion
    if (from === to) {
      return { rate: 1.0, source: 'direct' };
    }

    try {
      // Check cache first
      const cachedRate = await this.getCachedRate(from, to, dateStr);
      if (cachedRate) {
        console.log(`[CurrencyConverter] Using cached rate for ${from}→${to} on ${dateStr}: ${cachedRate.exchange_rate}`);
        return { rate: cachedRate.exchange_rate, source: 'cache' };
      }

      // Fetch from ExchangeRate.host with retry logic
      const rate = await this.fetchRateWithRetry(from, to, dateStr);
      
      if (rate !== null) {
        // Cache the successful result
        await this.cacheRate(from, to, dateStr, rate, 'exchangerate.host');
        console.log(`[CurrencyConverter] Fetched and cached rate for ${from}→${to} on ${dateStr}: ${rate}`);
        return { rate, source: 'api' };
      }

      return null;

    } catch (error) {
      console.error(`[CurrencyConverter] Error getting historical rate for ${from}→${to} on ${dateStr}:`, error);
      return null;
    }
  }

  /**
   * Convert amount to USD using historical rates
   */
  async convertToUSD(
    amount: number, 
    currency: string, 
    transactionDate: Date
  ): Promise<ConversionResult> {
    try {
      // Reject only zero values - negative values are valid (refunds, reimbursements, corrections)
      if (amount === 0) {
        return { 
          usd_amount: null, 
          exchange_rate: null, 
          success: false, 
          error: 'Invalid amount: cannot be zero' 
        };
      }

      // Preserve sign for negative amounts (refunds, loan repayments, corrections)
      const isNegative = amount < 0;
      const absoluteAmount = Math.abs(amount);

      const currencyCode = currency.toUpperCase();
      const dateStr = transactionDate.toISOString().split('T')[0];

      // Already USD - preserve the sign for negative values
      if (currencyCode === 'USD') {
        return { 
          usd_amount: amount, // Preserves sign
          exchange_rate: 1.0, 
          success: true,
          source: 'direct',
          conversion_date: dateStr
        };
      }

      // Check if currency is supported
      if (!(await this.isCurrencySupported(currencyCode))) {
        return { 
          usd_amount: null, 
          exchange_rate: null, 
          success: false, 
          error: `Currency ${currencyCode} is not supported for conversion` 
        };
      }

      // Get exchange rate (using absolute value)
      const rateResult = await this.getHistoricalRate(currencyCode, 'USD', transactionDate);
      
      if (!rateResult) {
        return { 
          usd_amount: null, 
          exchange_rate: null, 
          success: false, 
          error: `No exchange rate available for ${currencyCode} on ${dateStr}` 
        };
      }

      // Calculate USD amount with proper rounding using absolute value
      const usdAmount = Math.round(absoluteAmount * rateResult.rate * 100) / 100;
      
      // Apply the original sign to the converted amount
      const finalAmount = isNegative ? -usdAmount : usdAmount;

      return {
        usd_amount: finalAmount,
        exchange_rate: rateResult.rate,
        success: true,
        source: rateResult.source,
        conversion_date: dateStr
      };

    } catch (error) {
      console.error('[CurrencyConverter] Conversion error:', error);
      return {
        usd_amount: null,
        exchange_rate: null,
        success: false,
        error: error instanceof Error ? error.message : 'Conversion failed'
      };
    }
  }

  /**
   * Convert a single transaction to USD
   */
  async convertTransaction(transactionId: string): Promise<{ success: boolean; error?: string }> {
    try {
      const supabase = getSupabaseAdmin();
      if (!supabase) {
        return { success: false, error: 'Database connection not available' };
      }

      // Get transaction details
      const { data: transaction, error: fetchError } = await supabase
        .from('transactions')
        .select('*')
        .eq('uuid', transactionId)
        .single();

      if (fetchError || !transaction) {
        return { success: false, error: 'Transaction not found' };
      }

      // Skip if already converted or is USD
      if (transaction.value_usd !== null || transaction.currency === 'USD') {
        return { success: true };
      }

      // Use value_date if available, otherwise transaction_date
      const conversionDate = new Date(transaction.value_date || transaction.transaction_date);

      // Convert to USD
      const result = await this.convertToUSD(
        transaction.value,
        transaction.currency,
        conversionDate
      );

      if (!result.success) {
        // Mark as unconvertible
        await supabase
          .from('transactions')
          .update({
            usd_convertible: false,
            usd_conversion_date: new Date().toISOString()
          })
          .eq('uuid', transactionId);

        return { success: false, error: result.error };
      }

      // Update transaction with USD value
      const { error: updateError } = await supabase
        .from('transactions')
        .update({
          value_usd: result.usd_amount,
          exchange_rate_used: result.exchange_rate,
          usd_conversion_date: new Date().toISOString(),
          usd_convertible: true
        })
        .eq('uuid', transactionId);

      if (updateError) {
        console.error('[CurrencyConverter] Error updating transaction:', updateError);
        return { success: false, error: updateError.message };
      }

      console.log(`[CurrencyConverter] Successfully converted transaction ${transactionId}: ${transaction.value} ${transaction.currency} → $${result.usd_amount} USD`);
      return { success: true };

    } catch (error) {
      console.error('[CurrencyConverter] Transaction conversion error:', error);
      return { 
        success: false, 
        error: error instanceof Error ? error.message : 'Unknown error' 
      };
    }
  }

  /**
   * Check if a currency is supported for conversion
   */
  async isCurrencySupported(currencyCode: string): Promise<boolean> {
    const code = currencyCode.toUpperCase();
    
    if (code === 'USD') return true;

    try {
      const supabase = getSupabaseAdmin();
      if (!supabase) {
        // Fallback to default list if no database
        return this.DEFAULT_CURRENCIES.includes(code);
      }

      const { data, error } = await supabase
        .from('supported_currencies')
        .select('is_supported')
        .eq('code', code)
        .single();

      if (error || !data) {
        // Check if it's in our default list
        return this.DEFAULT_CURRENCIES.includes(code);
      }

      return data.is_supported;
    } catch (error) {
      console.warn(`[CurrencyConverter] Error checking currency support for ${code}:`, error);
      return this.DEFAULT_CURRENCIES.includes(code);
    }
  }

  /**
   * Get list of supported currencies
   */
  async getSupportedCurrencies(refresh = false): Promise<SupportedCurrency[]> {
    try {
      const supabase = getSupabaseAdmin();
      if (!supabase) {
        // Return default currencies if no database
        return this.DEFAULT_CURRENCIES.map(code => ({
          code,
          name: this.getCurrencyName(code),
          is_supported: true
        }));
      }

      if (!refresh) {
        // Try to get from database first
        const { data: dbCurrencies, error } = await supabase
          .from('supported_currencies')
          .select('*')
          .eq('is_supported', true)
          .order('code');

        if (!error && dbCurrencies && dbCurrencies.length > 0) {
          return dbCurrencies;
        }
      }

      // Fetch supported currencies from ExchangeRate.host
      const supportedCurrencies = await this.fetchSupportedCurrencies();
      
      // Update database
      await this.updateSupportedCurrencies(supportedCurrencies);

      // Return updated list
      const { data: updatedCurrencies } = await supabase
        .from('supported_currencies')
        .select('*')
        .eq('is_supported', true)
        .order('code');

      return updatedCurrencies || [];

    } catch (error) {
      console.error('[CurrencyConverter] Error getting supported currencies:', error);
      
      // Return fallback currencies
      return this.DEFAULT_CURRENCIES.map(code => ({
        code,
        name: this.getCurrencyName(code),
        is_supported: true
      }));
    }
  }

  /**
   * Bulk convert multiple transactions
   */
  async bulkConvertTransactions(transactionIds: string[]): Promise<{
    success: number;
    failed: number;
    errors: Array<{ id: string; error: string }>;
  }> {
    const results = {
      success: 0,
      failed: 0,
      errors: [] as Array<{ id: string; error: string }>
    };

    console.log(`[CurrencyConverter] Starting bulk conversion of ${transactionIds.length} transactions`);

    for (const id of transactionIds) {
      try {
        const result = await this.convertTransaction(id);
        if (result.success) {
          results.success++;
        } else {
          results.failed++;
          results.errors.push({ id, error: result.error || 'Unknown error' });
        }
      } catch (error) {
        results.failed++;
        results.errors.push({ 
          id, 
          error: error instanceof Error ? error.message : 'Unexpected error' 
        });
      }
    }

    console.log(`[CurrencyConverter] Bulk conversion completed: ${results.success} success, ${results.failed} failed`);
    return results;
  }

  /**
   * Get conversion statistics
   */
  async getConversionStats(activityId?: string): Promise<ConversionStats> {
    try {
      const supabase = getSupabaseAdmin();
      if (!supabase) {
        return this.getEmptyStats();
      }

      let query = supabase.from('transactions').select('*');
      
      if (activityId) {
        query = query.eq('activity_id', activityId);
      }

      const { data: transactions, error } = await query;

      if (error || !transactions) {
        console.error('[CurrencyConverter] Error fetching transactions for stats:', error);
        return this.getEmptyStats();
      }

      const stats = {
        total_transactions: transactions.length,
        converted_transactions: 0,
        unconvertible_transactions: 0,
        pending_transactions: 0,
        usd_transactions: 0,
        conversion_rate: 0
      };

      transactions.forEach((txn: any) => {
        if (txn.currency === 'USD') {
          stats.usd_transactions++;
        } else if (txn.value_usd !== null) {
          stats.converted_transactions++;
        } else if (txn.usd_convertible === false) {
          stats.unconvertible_transactions++;
        } else {
          stats.pending_transactions++;
        }
      });

      const convertibleTotal = stats.total_transactions - stats.usd_transactions - stats.unconvertible_transactions;
      if (convertibleTotal > 0) {
        stats.conversion_rate = Math.round((stats.converted_transactions / convertibleTotal) * 100 * 100) / 100;
      }

      return stats;

    } catch (error) {
      console.error('[CurrencyConverter] Error getting conversion stats:', error);
      return this.getEmptyStats();
    }
  }

  /**
   * Private helper methods
   */
  private async getCachedRate(from: string, to: string, date: string): Promise<ExchangeRate | null> {
    try {
      const supabase = getSupabaseAdmin();
      if (!supabase) return null;

      const { data, error } = await supabase
        .from('exchange_rates')
        .select('*')
        .eq('from_currency', from)
        .eq('to_currency', to)
        .eq('rate_date', date)
        .single();

      if (error || !data) return null;

      // Check if cache is still valid
      const cacheAge = Date.now() - new Date(data.created_at || 0).getTime();
      if (cacheAge > this.CACHE_DURATION) {
        return null;
      }

      return data;
    } catch (error) {
      return null;
    }
  }

  private async cacheRate(from: string, to: string, date: string, rate: number, source: string): Promise<void> {
    try {
      const supabase = getSupabaseAdmin();
      if (!supabase) return;

      await supabase
        .from('exchange_rates')
        .upsert({
          from_currency: from,
          to_currency: to,
          exchange_rate: rate,
          rate_date: date,
          source,
          created_at: new Date().toISOString()
        }, {
          onConflict: 'from_currency,to_currency,rate_date'
        });
    } catch (error) {
      console.warn('[CurrencyConverter] Failed to cache rate:', error);
    }
  }

  private async fetchRateWithRetry(from: string, to: string, date: string): Promise<number | null> {
    for (let attempt = 1; attempt <= this.MAX_RETRIES; attempt++) {
      try {
        const rate = await this.fetchRateFromAPI(from, to, date);
        if (rate !== null) return rate;
      } catch (error) {
        console.warn(`[CurrencyConverter] Attempt ${attempt} failed for ${from}→${to} on ${date}:`, error);
        
        if (attempt < this.MAX_RETRIES) {
          await new Promise(resolve => setTimeout(resolve, this.RETRY_DELAY * attempt));
        }
      }
    }
    return null;
  }

  private async fetchRateFromAPI(from: string, to: string, date: string): Promise<number | null> {
    try {
      // Validate date is not in the future
      const today = new Date().toISOString().split('T')[0];
      if (date > today) {
        console.warn(`[CurrencyConverter] Cannot fetch rate for future date: ${date}`);
        return null;
      }

      // Use ExchangeRate.host historical endpoint
      const url = `${this.API_BASE_URL}/${date}?base=${from}&symbols=${to}`;
      console.log(`[CurrencyConverter] Fetching rate from: ${url}`);
      
      const response = await fetch(url, {
        headers: {
          'User-Agent': 'AIMS Currency Converter/1.0'
        }
      });

      if (!response.ok) {
        throw new Error(`API request failed: ${response.status} ${response.statusText}`);
      }

      const data = await response.json();
      
      // Check for API errors
      if (!data.success && data.error) {
        throw new Error(`API error: ${data.error.info || data.error}`);
      }

      // Extract rate
      const rates = data.rates || {};
      if (!(to in rates)) {
        throw new Error(`Rate for ${to} not found in response`);
      }

      const rate = rates[to];
      if (typeof rate !== 'number' || isNaN(rate) || rate <= 0) {
        throw new Error(`Invalid rate value: ${rate}`);
      }

      return rate;

    } catch (error) {
      console.error(`[CurrencyConverter] API fetch error for ${from}→${to} on ${date}:`, error);
      throw error;
    }
  }

  private async fetchSupportedCurrencies(): Promise<string[]> {
    try {
      const url = `${this.API_BASE_URL}/symbols`;
      const response = await fetch(url);
      
      if (!response.ok) {
        throw new Error(`Failed to fetch supported currencies: ${response.status}`);
      }

      const data = await response.json();
      const symbols = data.symbols || {};
      
      return Object.keys(symbols);
    } catch (error) {
      console.warn('[CurrencyConverter] Failed to fetch supported currencies from API, using defaults:', error);
      return this.DEFAULT_CURRENCIES;
    }
  }

  private async updateSupportedCurrencies(currencyCodes: string[]): Promise<void> {
    try {
      const supabase = getSupabaseAdmin();
      if (!supabase) return;

      // Mark all as unsupported first
      await supabase
        .from('supported_currencies')
        .update({ is_supported: false, last_checked: new Date().toISOString() });

      // Update/insert supported currencies
      for (const code of currencyCodes) {
        await supabase
          .from('supported_currencies')
          .upsert({
            code: code.toUpperCase(),
            name: this.getCurrencyName(code),
            is_supported: true,
            last_checked: new Date().toISOString()
          }, {
            onConflict: 'code'
          });
      }
    } catch (error) {
      console.error('[CurrencyConverter] Error updating supported currencies:', error);
    }
  }

  private getCurrencyName(code: string): string {
    const names: Record<string, string> = {
      'USD': 'US Dollar', 'EUR': 'Euro', 'GBP': 'British Pound', 'JPY': 'Japanese Yen',
      'AUD': 'Australian Dollar', 'CAD': 'Canadian Dollar', 'CHF': 'Swiss Franc', 'CNY': 'Chinese Yuan',
      'SEK': 'Swedish Krona', 'NZD': 'New Zealand Dollar', 'MXN': 'Mexican Peso', 'SGD': 'Singapore Dollar',
      'HKD': 'Hong Kong Dollar', 'NOK': 'Norwegian Krone', 'TRY': 'Turkish Lira', 'RUB': 'Russian Ruble',
      'INR': 'Indian Rupee', 'BRL': 'Brazilian Real', 'ZAR': 'South African Rand', 'KRW': 'South Korean Won',
      'PLN': 'Polish Zloty', 'CZK': 'Czech Koruna', 'HUF': 'Hungarian Forint', 'RON': 'Romanian Leu',
      'BGN': 'Bulgarian Lev', 'HRK': 'Croatian Kuna', 'DKK': 'Danish Krone', 'THB': 'Thai Baht',
      'MYR': 'Malaysian Ringgit', 'PHP': 'Philippine Peso', 'IDR': 'Indonesian Rupiah', 'VND': 'Vietnamese Dong',
      'CLP': 'Chilean Peso', 'PEN': 'Peruvian Sol', 'COP': 'Colombian Peso', 'ARS': 'Argentine Peso',
      'LKR': 'Sri Lankan Rupee', 'PKR': 'Pakistani Rupee', 'BDT': 'Bangladeshi Taka', 'NPR': 'Nepalese Rupee',
      'AFN': 'Afghan Afghani', 'MMK': 'Myanmar Kyat', 'KHR': 'Cambodian Riel', 'LAK': 'Lao Kip'
    };
    return names[code.toUpperCase()] || `${code.toUpperCase()} Currency`;
  }

  private getEmptyStats(): ConversionStats {
    return {
      total_transactions: 0,
      converted_transactions: 0,
      unconvertible_transactions: 0,
      pending_transactions: 0,
      usd_transactions: 0,
      conversion_rate: 0
    };
  }
}

// Export singleton instance
export const currencyConverter = new EnhancedCurrencyConverter();
export default currencyConverter;